filter-nm/ftnoir_filter_nm.cpp

```diff
@@ -0,0 +1,48 @@
+/* Copyright (c) 2023 Tom Brazier <tom_github@firstsolo.net>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ */
+#include "ftnoir_filter_nm.h"
+#include "compat/math-imports.hpp"
+#include "compat/macros.h"
+
+#include "api/plugin-api.hpp"
+#include "opentrack/defs.hpp"
+
+#include <algorithm>
+
+filter_nm::filter_nm()
+{
+}
+
+void filter_nm::filter(const double* input, double* output)
+{
+    // order of axes: x, y, z, yaw, pitch, roll
+    if (unlikely(first_run))
+    {
+        first_run = false;
+        t.start();
+
+        std::fill(speeds, speeds + 6, 0.0);
+        std::copy(input, input + 6, filtered_output);
+    }
+    else
+    {
+        const double dt = t.elapsed_seconds();
+        t.start();
+
+        for (int i = 0; i < 6; i++)
+        {
+            double speed = (input[i] - last_input[i]) / dt;
+            speeds[i] += dt * (double)s.responsiveness[i] * (speed - speeds[i]);
+            filtered_output[i] += dt * (double)s.responsiveness[i] * min(1.0, abs(speeds[i]) / (double)s.drift_speeds[i]) * (input[i] - filtered_output[i]);
```

Comment:
Both, in line 39 and line 40, when `dt * responsiveness` becomes large, the output would overshoot. Possibly it could become unstable and explode. Not sure.
I recognize you have two EWA going on there. So consider limiting the smoothing factors `alpha`, e.g. shown for the speed here, by
```
timescale = 1./responsiveness;
alpha = dt / (dt + timescale);
speeds[i] += alpha * (speed - speeds[i]); // EWA
```
In line 40, I think it would be fine then to just replace `dt * (double)s.responsiveness[i]` by `alpha`.

Reply:
Good point. In my head dt was always going to be very small but at a max value of 20 for `responsiveness` and below 20Hz camera frame rate that's going to be a problem. Your suggested EWA approach is spot on. I'll do that.